burstphoto/align.metal

```diff
@@ -416,3 +740,992 @@ kernel void compute_merge_weight(texture2d<float, access::read> texture_diff [[t
+    float4 const norm_factor = float4(float(n_textures*tile_size*tile_size), float(n_textures*tile_size*tile_size), float(n_textures*tile_size*tile_size), float(n_textures*tile_size*tile_size));
+       
+    float coefRe, coefIm;
+    float4 Re, Im, dataRe, dataIm;
+    
+    // row-wise one-dimensional discrete Fourier transform along x-direction
+    for (int dn = 0; dn < tile_size; dn++) {
+        for (int dm = 0; dm < tile_size; dm++) {
+             
+            int const m = 2*(m0 + dm);
+            int const n = n0 + dn;
+            
+            // fill with zeros
+            Re = zeros;
+            Im = zeros;
+            
+            for (int dx = 0; dx < tile_size; dx++) {
+                                  
+                int const x = 2*(m0 + dx);
+              
+                // calculate coefficients
+                coefRe = cos(angle*dm*dx);
+                coefIm = sin(angle*dm*dx);
+                
+                dataRe = in_texture_ft.read(uint2(x+0, n));
+                dataIm = in_texture_ft.read(uint2(x+1, n));
+                
+                Re += (coefRe*dataRe - coefIm*dataIm);
+                Im += (coefIm*dataRe + coefRe*dataIm);
+            }
+            
+            // write into temporary textures
+            tmp_texture_ft.write(Re, uint2(m+0, n));
+            tmp_texture_ft.write(Im, uint2(m+1, n));
+        }
+    }
+    
+    // column-wise one-dimensional discrete Fourier transform along y-direction
+    for (int dm = 0; dm < tile_size; dm++) {
+        for (int dn = 0; dn < tile_size; dn++) {
+              
+            int const m = m0 + dm;
+            int const n = n0 + dn;
+             
+            // fill with zeros
+            Re = zeros;
+              
+            for (int dy = 0; dy < tile_size; dy++) {
+                                  
+                int const y = n0 + dy;
+                
+                // calculate coefficients
+                coefRe = cos(angle*dn*dy);
+                coefIm = sin(angle*dn*dy);
+                           
+                dataRe = tmp_texture_ft.read(uint2(2*m+0, y));
+                dataIm = tmp_texture_ft.read(uint2(2*m+1, y));
+                            
+                Re += (coefRe*dataRe - coefIm*dataIm);
+            }
+            
+            // normalize result
+            Re = Re/norm_factor;
+            out_texture.write(Re, uint2(m, n));
+        }
+    }
+}
+
+
+// This is the most important function required for the frequency-based merging approach. It is based on ideas from several publications:
+// [Hasinoff 2016]: https://graphics.stanford.edu/papers/hdrp/hasinoff-hdrplus-sigasia16.pdf
+// [Monod 2021]: https://www.ipol.im/pub/art/2021/336/
+// [Liba 2019]: https://graphics.stanford.edu/papers/night-sight-sigasia19/night-sight-sigasia19.pdf
+// [Delbracio 2015]: https://openaccess.thecvf.com/content_cvpr_2015/papers/Delbracio_Burst_Deblurring_Removing_2015_CVPR_paper.pdf
```

Comment:
While I think it's great that the references used are included, I think it would be more useful to explicitly state what was used from each of them. E.g. the deblurring was from Delbracio.